File: algorithmic-toolbox/src/week2/scala/MinimumDotProduct.scala
object MinimumDotProduct {

  def main(args: Array[String]) = {
    val in = new java.util.Scanner(System.in)
    val n = in.nextInt
    println((1 to n)
      .map(n => in.nextLong)
      .sorted
      .zip(
        (1 to n)
          .map(n => in.nextLong)
          .sortWith(_ > _))
      .map(n => n._1 * n._2)
      .sum)
  }
}

File: algorithmic-toolbox/src/week2/java/MinimumDotProduct.java
package week2.java;
import java.util.Arrays;
import java.util.Scanner;

class MinimumDotProduct {

    public static void main(final String[] args) {
        try (final Scanner in = new Scanner(System.in)) {
            final int n = in.nextInt();
            final int[] a = readArray(in, n), b = readArray(in, n);

            long result = 0;
            for (int i = 0; i < n; i++) {
                result += (1L * a[n-i-1] * b[i]);
            }

            System.out.println(result);
        }
    }

    private static int[] readArray(final Scanner in, final int length) {
        final int[] result = new int[length];
        for (int i = 0; i < length; i++) {
            result[i] = in.nextInt();
        }

        Arrays.sort(result);
        return result;
    }
}